---
title: "reCAT_run"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load some packages

```{r load, warning=FALSE}
suppressPackageStartupMessages(library(SummarizedExperiment))
suppressPackageStartupMessages(library(MultiAssayExperiment))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(doParallel))
suppressPackageStartupMessages(library(mclust))
suppressPackageStartupMessages(library(cluster))
suppressPackageStartupMessages(library(TSP))
```

Load the data 

```{r data}
# load("reCAT_master/data/ola_mES_2i.RData")
cc_data <- readRDS("embryonic_data/EMTAB2805.rds")

# Get tpms at gene level, TPM = (Transcripts Per Kilobase Million), TPM is generated beforehand through
# 1. Divide the read counts by the length of each gene in kilobases. This gives you reads per kilobase (RPK).
# 2. Count up all the RPK values in a sample and divide this number by 1,000,000. This is your “per million” scaling factor.
# 3. Divide the RPK values by the “per million” scaling factor. This gives you TPM.

# Assays is a subclass of RangedSummarizedExperiment in which A list or SimpleList of matrix-like elements, or a matrix-like object. 
# All elements of the list must have the same dimensions, and dimension names (if present) must be consistent across elements and 
# with the row names of rowRanges and colData


# Retrieves the gene vs. counts data
cc_counts <- assays(cc_data)$gene

# remove spike-ins from expression table,  ERCC = External RNA Controls Consortium (ERCC)
# External RNA-controls to check for variability, ie: not needed for the analysis
ercc <- grep("ERCC",rownames(cc_counts))
cc_counts <- cc_counts[-ercc,]

# remove the .1 etc from the ensembl gene name
gname <- rownames(cc_counts)
ensname <- gsub("\\.\\d+","",gname)
rownames(cc_counts) <- ensname

# Transposes the matrix to work with reCAT's functions
cc_counts <- t(cc_counts)

# Removes all rows with NA values from the matrix
cc_counts <- na.omit(cc_counts)

# Inspect the matrix
indices <- c(1, 20, 33, 44, 64)
cc_counts[indices,indices]
```

```{r preprocess}
source("reCAT_master/R/get_test_exp.R")
pre_counts <- cc_counts
pre_counts <- get_test_exp(pre_counts)
```

```{r order}
source("reCAT_master/R/get_ordIndex.R")
ordIndex <- get_ordIndex(cc_counts, 1)
```

Get cell-cycle scores

```{r scores}
source("reCAT_master/R/get_score.R")
score_result <- get_score(t(cc_counts))
score_result$bayes_score
score_result$mean_score
```