---
title: "Detecting cell-cycle stage difference in Buettner mESC cells"
author: "Donghyung Lee, minor modifications by Erik Hedman"
date: "`r Sys.Date()`"
output:
  html_document: default
---

**Clarification statement**. In this markdown script the parts I have modified are stated with my name, Erik. Some code blocks are provided by Johan Reimegård, my client. The original comments and code blocks are mostly stated as "we", and are then referred to Donghyung Lee and his team. 

Here, we illustrate how to use the iasva package to detect cell cycle stage
difference within single cell RNA sequencing data. Erik then extracted gene lists created with iasva and compared these lists with gene lists used in [Cyclone](https://www.sciencedirect.com/science/article/pii/S1046202315300098) and [Seurat](https://www.nature.com/articles/nbt.4096). Erik used single cell RNA
sequencing (scRNA-Seq) data obtained from Mus musculus mESC samples
([Scialdone *et al.* 2015](https://www.ncbi.nlm.nih.gov/pubmed/25599176)).
Other human glioblastoma samples can be found in a R data package ("iasvaExamples") containing data examples for IA-SVA (https://github.com/dleelab/iasvaExamples). 
To install the package, follow the instruction provided in the GitHub page.

## Install packages

```{r install_packages, echo=TRUE, eval=FALSE}
#devtools
library(devtools)
#iasva
devtools::install_github("UcarLab/iasva")
#iasvaExamples  
devtools::install_github("dleelab/iasvaExamples")
```

## Load packages

Probably, not all packages are needed. Erik has not looked into this.

```{r load_packages, echo=TRUE, message=FALSE}
rm(list=ls())
library(irlba) # partial SVD, the augmented implicitly restarted Lanczos bidiagonalization algorithm
library(iasva)
library(iasvaExamples)
library(sva)
library(SCnorm)
library(Seurat)
library(scran)
library(scater)
library(Rtsne)
library(pheatmap)
library(corrplot)
library(DescTools) #pcc i.e., Pearson's contingency coefficient
library(RColorBrewer)
library(SummarizedExperiment)
library(MultiAssayExperiment)
library(vioplot)
color.vec <- brewer.pal(3, "Set1")

# Normalization.
normalize <- function(counts) 
{
    normfactor <- colSums(counts)
    return(t(t(counts)/normfactor)*median(normfactor))
}
```

## Load the Buettner single cell RNA-Seq data

From the Conquer database (http://imlspenticton.uzh.ch:3838/conquer/) the Buettner dataset was fetched:

1	EMTAB2805 (PMID 25599176)	Buettner2015 Mus musculus	mESC in different cell cycle stages	288	SMARTer C1	full-length

This code block was provided by the client Johan Reimegård.

```{r load_data, echo=TRUE}
buettner.data <- readRDS("/Users/erikh/Desktop/git/scrna-cell-cycle/embryonic_data/EMTAB2805.rds")

# get tpms at gene level
buet_exprs <- assays(experiments(buettner.data)[["gene"]])[["TPM"]]
dim(buet_exprs)
# remove the .1 etc from the ensembl gene name
gname <- rownames(buet_exprs)
ensname <- gsub("\\.\\d+","",gname)
rownames(buet_exprs) <- ensname

# remove spike-ins from expression table
ercc <- grep("ERCC",rownames(buet_exprs))
buet_exprs <- buet_exprs[-ercc,]

# metadata with cell cycle stage assignment
M <- as.data.frame(colData(buettner.data))
head(M)
dim(buet_exprs)
```

## Filter out clearly low quality cells from Buettner data

Clearly some outlier cells on negative PC1 with low number of genes with pca using full dataset.
Mainly S-phase cells, even if most S-phase cells have a high median expression.

This code block was provided by the client Johan Reimegård.

```{r filter}
# filter out cells with low nDet
nDet <- colSums(buet_exprs>1)
hist(nDet,n=100)
# make a cutoff at 4000 genes, removes 16 cells
keep <- which(nDet>4000)

buet_exprs<-buet_exprs[,keep]
M <- M[keep,]
dim(buet_exprs)

```

## Normalization for Buettner data

This code block was provided by the client Johan Reimegård.

```{r normalization, fig.width=8, fig.height=6}
## count-depth relationship for all genes
norm_buet <- SingleCellExperiment(assays=list(counts = buet_exprs, logcounts = log2(buet_exprs+1)))
dim(norm_buet)
```

## Calculate the number of detected genes 

It is well known that the number of detected genes in each cell explains
a very large portion of variability in scRNA-Seq data 
([Hicks et. al. 2015 BioRxiv](http://biorxiv.org/content/early/2015/08/25/025528),
[McDavid et. al. 2016 Nature Biotechnology](http://www.nature.com/nbt/journal/v34/n6/full/nbt.3498.html)).
Frequently, the first principal component of log-transformed scRNA-Seq read counts is highly correlated with the number of detected genes (e.g., r > 0.9). Here, Erik calculates the number of detected genes for Mus musculus mESC cells, which will be used as an known factor in the IA-SVA analyses. The code is not modified by Erik, only expression matrix differ from original script.

```{r num_detected_genes, echo=TRUE, fig.width=7, fig.height=4}
Num_Detected_Genes_buet <- colSums(buet_exprs>0)
Geo_Lib_buet <- colSums(log(buet_exprs+1))
summary(Geo_Lib_buet)
barplot(Geo_Lib_buet, xlab="Cell", las=2, ylab = "Geometric library size")
lcounts_buet <- log(buet_exprs + 1)
# PC1 and Geometric library size correlation
pc1_buet = irlba(lcounts_buet - rowMeans(lcounts_buet), 1)$v[,1] ## partial SVD
cor(Num_Detected_Genes_buet, pc1_buet)
cor(Geo_Lib_buet, pc1_buet)
```

## Run IA-SVA

Here, we run IA-SVA using Geo_Lib_Size as a known factor. Erik identifies seven hidden factors. SVs are plotted in a pairwise fashion to uncover which SVs can seperate cells. Lastly, the correlation plot of all SVs is plotted. 

```{r run_iasva, echo=TRUE, fig.width= 7, fig.height=6}
set.seed(3445)
mod_buet <- model.matrix(~Geo_Lib_buet)
summ_exp_buet <- SummarizedExperiment(assays = buet_exprs)
iasva.res_buet <- iasva(summ_exp_buet, as.matrix(mod_buet[,-1]),verbose=FALSE, permute=FALSE, num.sv=7)
iasva.sv_buet <- iasva.res_buet$sv

Cluster_buet_7 <- as.factor(iasva.sv_buet[,7] < 0.1) 
levels(Cluster_buet_7)=c("Cluster1","Cluster2")
table(Cluster_buet_7)

pairs(iasva.sv_buet[,1:7], main="IA-SVA_7", pch=21, col=color.vec[Cluster_buet_7],
      bg=color.vec[Cluster_buet_7], oma=c(4,4,6,14))
legend("right", levels(Cluster_buet_7), fill=color.vec, bty="n")

plot(iasva.sv_buet[,c(1,7)], main="IA-SVA", pch=21, xlab="SV1", ylab="SV7",
     col=color.vec[Cluster_buet_7], bg=color.vec[Cluster_buet_7])

Cluster_buet_6 <- as.factor(iasva.sv_buet[,6] < 0.1) 
levels(Cluster_buet_6)=c("Cluster1","Cluster2")
table(Cluster_buet_6)

pairs(iasva.sv_buet[,1:7], main="IA-SVA_6", pch=21, col=color.vec[Cluster_buet_6],
      bg=color.vec[Cluster_buet_6], oma=c(4,4,6,14))
legend("right", levels(Cluster_buet_6), fill=color.vec, bty="n")

plot(iasva.sv_buet[,c(1,6)], main="IA-SVA", pch=21, xlab="SV1", ylab="SV6",
     col=color.vec[Cluster_buet_6], bg=color.vec[Cluster_buet_6])

Cluster_buet_5 <- as.factor(iasva.sv_buet[,5] < 0.1) 
levels(Cluster_buet_5)=c("Cluster1","Cluster2")
table(Cluster_buet_5)

pairs(iasva.sv_buet[,1:7], main="IA-SVA_5", pch=21, col=color.vec[Cluster_buet_5],
      bg=color.vec[Cluster_buet_5], oma=c(4,4,6,14))
legend("right", levels(Cluster_buet_5), fill=color.vec, bty="n")

plot(iasva.sv_buet[,c(1,5)], main="IA-SVA", pch=21, xlab="SV1", ylab="SV5",
     col=color.vec[Cluster_buet_5], bg=color.vec[Cluster_buet_5])

Cluster_buet_4 <- as.factor(iasva.sv_buet[,4] < 0.1) 
levels(Cluster_buet_4)=c("Cluster1","Cluster2")
table(Cluster_buet_4)

pairs(iasva.sv_buet[,1:7], main="IA-SVA_4", pch=21, col=color.vec[Cluster_buet_4],
      bg=color.vec[Cluster_buet_4], oma=c(4,4,6,14))
legend("right", levels(Cluster_buet_4), fill=color.vec, bty="n")

plot(iasva.sv_buet[,c(1,4)], main="IA-SVA", pch=21, xlab="SV1", ylab="SV4",
     col=color.vec[Cluster_buet_4], bg=color.vec[Cluster_buet_4])

Cluster_buet_3 <- as.factor(iasva.sv_buet[,3] < 0.1) 
levels(Cluster_buet_3)=c("Cluster1","Cluster2")
table(Cluster_buet_3)

pairs(iasva.sv_buet[,1:7], main="IA-SVA_3", pch=21, col=color.vec[Cluster_buet_3],
      bg=color.vec[Cluster_buet_3], oma=c(4,4,6,14))
legend("right", levels(Cluster_buet_3), fill=color.vec, bty="n")

plot(iasva.sv_buet[,c(1,3)], main="IA-SVA", pch=21, xlab="SV1", ylab="SV3",
     col=color.vec[Cluster_buet_3], bg=color.vec[Cluster_buet_3])

Cluster_buet_2 <- as.factor(iasva.sv_buet[,2] < 0.1) 
levels(Cluster_buet_2)=c("Cluster1","Cluster2")
table(Cluster_buet_2)

pairs(iasva.sv_buet[,1:7], main="IA-SVA_2", pch=21, col=color.vec[Cluster_buet_2],
      bg=color.vec[Cluster_buet_2], oma=c(4,4,6,14))
legend("right", levels(Cluster_buet_2), fill=color.vec, bty="n")

plot(iasva.sv_buet[,c(1,2)], main="IA-SVA", pch=21, xlab="SV1", ylab="SV2",
     col=color.vec[Cluster_buet_2], bg=color.vec[Cluster_buet_2])

Cluster_buet_1 <- as.factor(iasva.sv_buet[,1] < 0.1) 
levels(Cluster_buet_1)=c("Cluster1","Cluster2")
table(Cluster_buet_1)

pairs(iasva.sv_buet[,1:7], main="IA-SVA_1", pch=21, col=color.vec[Cluster_buet_1],
      bg=color.vec[Cluster_buet_1], oma=c(4,4,6,14))
legend("right", levels(Cluster_buet_1), fill=color.vec, bty="n")

plot(iasva.sv_buet[,c(1,2)], main="IA-SVA", pch=21, xlab="SV1", ylab="SV2",
     col=color.vec[Cluster_buet_1], bg=color.vec[Cluster_buet_1])

cor(Num_Detected_Genes_buet, iasva.sv_buet[,1:7])
cor(Geo_Lib_buet, iasva.sv_buet[,1:7])
corrplot(cor(iasva.sv_buet))
```

## Cell type assignment using scran R package

Here cyclone is used to assign cell cycle phases. The correlation plots explain how correlated each cell cycle phase is for each SV. One plot for G1, one for S, and one for G2M. Vioplot visualizes how well the different SVs separate the cell cycle phases all together. 

```{r run_scran, echo=TRUE, fig.width=6, fig.height=4}
mmu.pairs <- readRDS(system.file("exdata", 
                                "mouse_cycle_markers.rds", package="scran"))
assigned_buet <- cyclone(norm_buet, pairs=mmu.pairs)
head(assigned_buet$scores)
table(assigned_buet$phases)

G1_buet_SV <- iasva.sv_buet[assigned_buet$phases=="G1",]
S_buet_SV <- iasva.sv_buet[assigned_buet$phases=="S",]
G2M_buet_SV <- iasva.sv_buet[assigned_buet$phases=="G2M",]

corrplot(cor(G1_buet_SV))
corrplot(cor(S_buet_SV))
corrplot(cor(G2M_buet_SV))

vioplot(G1_buet_SV[,c(1)], S_buet_SV[,c(1)], G2M_buet_SV[,c(1)], names=c("G1", "S", "G2M"), col="gold")
title(xlab="Cell-cycle stage predictions", ylab="IA-SVA factor (SV1)")
vioplot(G1_buet_SV[,c(2)], S_buet_SV[,c(2)], G2M_buet_SV[,c(2)], names=c("G1", "S", "G2M"), 
   col="gold")
title(xlab="Cell-cycle stage predictions", ylab="IA-SVA factor (SV2)")
vioplot(G1_buet_SV[,c(3)], S_buet_SV[,c(3)], G2M_buet_SV[,c(3)], names=c("G1", "S", "G2M"), 
   col="gold")
title(xlab="Cell-cycle stage predictions", ylab="IA-SVA factor (SV3)")
vioplot(G1_buet_SV[,c(4)], S_buet_SV[,c(4)], G2M_buet_SV[,c(4)], names=c("G1", "S", "G2M"), 
   col="gold")
title(xlab="Cell-cycle stage predictions", ylab="IA-SVA factor (SV4)")
vioplot(G1_buet_SV[,c(5)], S_buet_SV[,c(5)], G2M_buet_SV[,c(5)], names=c("G1", "S", "G2M"), 
   col="gold")
title(xlab="Cell-cycle stage predictions", ylab="IA-SVA factor (SV5)")
vioplot(G1_buet_SV[,c(6)], S_buet_SV[,c(6)], G2M_buet_SV[,c(6)], names=c("G1", "S", "G2M"), 
   col="gold")
title(xlab="Cell-cycle stage predictions", ylab="IA-SVA factor (SV6)")
vioplot(G1_buet_SV[,c(7)], S_buet_SV[,c(7)], G2M_buet_SV[,c(7)], names=c("G1", "S", "G2M"), 
   col="gold")
title(xlab="Cell-cycle stage predictions", ylab="IA-SVA factor (SV7)")
```

## Kolmogorov–Smirnov test on the three cell cycle phases and the different SVs

The SV with the lowest KS value will be used in next section. The other SVs are ignored. The SV with lowest value should be the one best separating the cell cycle phases using the vioplot in previous section. Code need to be optimized.

```{r }
#KS-test between G1 and G2M
ks_G1_G2M_1 <- as.numeric(unlist(ks.test(G1_buet_SV[,1],G2M_buet_SV[,1])[2]))
ks_G1_G2M_2 <- as.numeric(unlist(ks.test(G1_buet_SV[,2],G2M_buet_SV[,2])[2]))
ks_G1_G2M_3 <- as.numeric(unlist(ks.test(G1_buet_SV[,3],G2M_buet_SV[,3])[2]))
ks_G1_G2M_4 <- as.numeric(unlist(ks.test(G1_buet_SV[,4],G2M_buet_SV[,4])[2]))
ks_G1_G2M_5 <- as.numeric(unlist(ks.test(G1_buet_SV[,5],G2M_buet_SV[,5])[2]))
ks_G1_G2M_6 <- as.numeric(unlist(ks.test(G1_buet_SV[,6],G2M_buet_SV[,6])[2]))
ks_G1_G2M_7 <- as.numeric(unlist(ks.test(G1_buet_SV[,7],G2M_buet_SV[,7])[2]))

#KS-test between G1 and S
ks_G1_S_1 <- as.numeric(unlist(ks.test(G1_buet_SV[,1],S_buet_SV[,1])[2]))
ks_G1_S_2 <- as.numeric(unlist(ks.test(G1_buet_SV[,2],S_buet_SV[,2])[2]))
ks_G1_S_3 <- as.numeric(unlist(ks.test(G1_buet_SV[,3],S_buet_SV[,3])[2]))
ks_G1_S_4 <- as.numeric(unlist(ks.test(G1_buet_SV[,4],S_buet_SV[,4])[2]))
ks_G1_S_5 <- as.numeric(unlist(ks.test(G1_buet_SV[,5],S_buet_SV[,5])[2]))
ks_G1_S_6 <- as.numeric(unlist(ks.test(G1_buet_SV[,6],S_buet_SV[,6])[2]))
ks_G1_S_7 <- as.numeric(unlist(ks.test(G1_buet_SV[,7],S_buet_SV[,7])[2]))

#KS-test between S and G2M
ks_G2M_S_1 <- as.numeric(unlist(ks.test(G2M_buet_SV[,1],S_buet_SV[,1])[2]))
ks_G2M_S_2 <- as.numeric(unlist(ks.test(G2M_buet_SV[,2],S_buet_SV[,2])[2]))
ks_G2M_S_3 <- as.numeric(unlist(ks.test(G2M_buet_SV[,3],S_buet_SV[,3])[2]))
ks_G2M_S_4 <- as.numeric(unlist(ks.test(G2M_buet_SV[,4],S_buet_SV[,4])[2]))
ks_G2M_S_5 <- as.numeric(unlist(ks.test(G2M_buet_SV[,5],S_buet_SV[,5])[2]))
ks_G2M_S_6 <- as.numeric(unlist(ks.test(G2M_buet_SV[,6],S_buet_SV[,6])[2]))
ks_G2M_S_7 <- as.numeric(unlist(ks.test(G2M_buet_SV[,7],S_buet_SV[,7])[2]))

#Multiply the different KS-tests for every SV to get a feeling of which SVs to exclude.
ks_SV1 <- ks_G1_G2M_1*ks_G1_S_1*ks_G2M_S_1
ks_SV2 <- ks_G1_G2M_2*ks_G1_S_2*ks_G2M_S_2
ks_SV3 <- ks_G1_G2M_3*ks_G1_S_3*ks_G2M_S_3
ks_SV4 <- ks_G1_G2M_4*ks_G1_S_4*ks_G2M_S_4
ks_SV5 <- ks_G1_G2M_5*ks_G1_S_5*ks_G2M_S_5
ks_SV6 <- ks_G1_G2M_6*ks_G1_S_6*ks_G2M_S_6
ks_SV7 <- ks_G1_G2M_7*ks_G1_S_7*ks_G2M_S_7
ks_vec <- data.frame(ks_SV1, ks_SV2, ks_SV3, ks_SV4, ks_SV5, ks_SV6, ks_SV7)
ks_vec
```

## Find marker genes for the detected heterogeneity (SV2).

Here, we use the find_markers() function. Erik finds marker genes that are significantly
associated with SV2, since this had the lowest KS-value. Default significance cutoff, and R-squared value > 0.3).

```{r find_markers_SV2, echo=TRUE, fig.width=6, fig.height=16}
# try different R2 thresholds
pdf(paste0("/Users/erikh/Desktop/git/scrna-cell-cycle/r_scripts/erik/ia_sva/sc_output/Clustering_analyses_figure3_sv2_buet.pdf"))
r2.results_buet <- study_R2(summ_exp_buet, iasva.sv_buet, selected.svs=2, no.clusters=2)
dev.off()

marker.counts.SV2_buet <- find_markers(summ_exp_buet, 
                            as.matrix(iasva.sv_buet[,c(2)]), rsq.cutoff = 0.35)
marker.counts.SV2.long_buet <- find_markers(summ_exp_buet, 
                              as.matrix(iasva.sv_buet[,c(2)]), rsq.cutoff = 0.3)
nrow(marker.counts.SV2_buet) 
nrow(marker.counts.SV2.long_buet)

anno.col2_buet <- data.frame(Cluster=Cluster_buet_2, SV1=iasva.sv_buet[,2])
rownames(anno.col2_buet) <- colnames(marker.counts.SV2_buet)
head(anno.col2_buet)

cluster.col_buet <- color.vec[1:2]
names(cluster.col_buet) <- as.vector(levels(Cluster_buet_2))
anno.colors_buet <- list(Cluster=cluster.col_buet)
anno.colors_buet

#heatmaps are excluded from the printed result due to irrelevance for this project.

#pheatmap(log(marker.counts.SV2_buet+1), show_colnames =FALSE, 
#         clustering_method = "ward.D2", cutree_cols = 2, annotation_col = anno.col2_buet,
#         annotation_colors = anno.colors_buet)

#pheatmap(log(marker.counts.SV2.long_buet+1), show_colnames =FALSE, 
#         clustering_method = "ward.D2", cutree_cols = 2, annotation_col = anno.col2_buet,
#         annotation_colors = anno.colors_buet)

gene.list_buet_SV2 <- rownames(marker.counts.SV2_buet)
write.table(gene.list_buet_SV2, file = paste0("/Users/erikh/Desktop/git/scrna-cell-cycle/r_scripts/erik/ia_sva/sc_output/CC_genes.short_buet_SV2.txt"),
            col.names =F, row.names = F, quote = F)

gene.list.long_buet_SV2 <- rownames(marker.counts.SV2.long_buet)
write.table(gene.list.long_buet_SV2, file = paste0("/Users/erikh/Desktop/git/scrna-cell-cycle/r_scripts/erik/ia_sva/sc_output/CC_genes.long_buet_SV2.txt"),
            col.names =F, row.names = F, quote = F)

```

## SV3

As a proof of concept SV3 is used to show the results with an SV with higher KS value.

```{r find_markers_SV3, echo=TRUE, fig.width=6, fig.height=16}
marker.counts.SV3_buet <- find_markers(summ_exp_buet, 
                            as.matrix(iasva.sv_buet[,c(3)]), rsq.cutoff = 0.35)
marker.counts.SV3.long_buet <- find_markers(summ_exp_buet, 
                              as.matrix(iasva.sv_buet[,c(3)]), rsq.cutoff = 0.3)
nrow(marker.counts.SV3_buet) 
nrow(marker.counts.SV3.long_buet)

anno.col2_buet <- data.frame(Cluster=Cluster_buet_3, SV1=iasva.sv_buet[,3])
rownames(anno.col2_buet) <- colnames(marker.counts.SV3_buet)
head(anno.col2_buet)

cluster.col_buet <- color.vec[1:2]
names(cluster.col_buet) <- as.vector(levels(Cluster_buet_3))
anno.colors_buet <- list(Cluster=cluster.col_buet)
anno.colors_buet

#heatmaps are excluded from the printed result due to irrelevance for this project.

#pheatmap(log(marker.counts.SV3_buet+1), show_colnames =FALSE, 
#         clustering_method = "ward.D2", cutree_cols = 2, annotation_col = anno.col2_buet,
#         annotation_colors = anno.colors_buet)

#pheatmap(log(marker.counts.SV3.long_buet+1), show_colnames =FALSE, 
#         clustering_method = "ward.D2", cutree_cols = 2, annotation_col = anno.col2_buet,
#         annotation_colors = anno.colors_buet)

gene.list_buet_SV3 <- rownames(marker.counts.SV3_buet)
write.table(gene.list_buet_SV3, file = paste0("/Users/erikh/Desktop/git/scrna-cell-cycle/r_scripts/erik/ia_sva/sc_output/CC_genes.short_buet_SV3.txt"),
            col.names =F, row.names = F, quote = F)

gene.list.long_buet_SV3 <- rownames(marker.counts.SV3.long_buet)
write.table(gene.list.long_buet_SV3, file = paste0("/Users/erikh/Desktop/git/scrna-cell-cycle/r_scripts/erik/ia_sva/sc_output/CC_genes.long_buet_SV3.txt"),
            col.names =F, row.names = F, quote = F)

```

## SV5

The SV with the second lowest KS value is tested to compare results with SV2.

```{r find_markers_SV5, echo=TRUE, fig.width=6, fig.height=16}
marker.counts.SV5_buet <- find_markers(summ_exp_buet, 
                            as.matrix(iasva.sv_buet[,c(5)]), rsq.cutoff = 0.35)
marker.counts.SV5.long_buet <- find_markers(summ_exp_buet, 
                              as.matrix(iasva.sv_buet[,c(5)]), rsq.cutoff = 0.3)
nrow(marker.counts.SV5_buet) 
nrow(marker.counts.SV5.long_buet)

anno.col2_buet <- data.frame(Cluster=Cluster_buet_5, SV1=iasva.sv_buet[,5])
rownames(anno.col2_buet) <- colnames(marker.counts.SV5_buet)
head(anno.col2_buet)

cluster.col_buet <- color.vec[1:2]
names(cluster.col_buet) <- as.vector(levels(Cluster_buet_5))
anno.colors_buet <- list(Cluster=cluster.col_buet)
anno.colors_buet

#pheatmap(log(marker.counts.SV5_buet+1), show_colnames =FALSE, 
#         clustering_method = "ward.D2", cutree_cols = 2, annotation_col = anno.col2_buet,
#         annotation_colors = anno.colors_buet)

#pheatmap(log(marker.counts.SV5.long_buet+1), show_colnames =FALSE, 
#         clustering_method = "ward.D2", cutree_cols = 2, annotation_col = anno.col2_buet,
#         annotation_colors = anno.colors_buet)

gene.list_buet_SV5 <- rownames(marker.counts.SV5_buet)
write.table(gene.list_buet_SV5, file = paste0("/Users/erikh/Desktop/git/scrna-cell-cycle/r_scripts/erik/ia_sva/sc_output/CC_genes.short_buet_SV5.txt"),
            col.names =F, row.names = F, quote = F)

gene.list.long_buet_SV5 <- rownames(marker.counts.SV5.long_buet)
write.table(gene.list.long_buet_SV5, file = paste0("/Users/erikh/Desktop/git/scrna-cell-cycle/r_scripts/erik/ia_sva/sc_output/CC_genes.long_buet_SV5.txt"),
            col.names =F, row.names = F, quote = F)

```

## Adding the Seurat gene list for comparison

This code block was provided by our client Johan Reimegård.

```{r convert_names}
savefile = "seurat_cc_genes_mouse.Rdata"
if (file.exists(savefile)){
  load(savefile)
}else{

# Load human ensembl attributes
human = biomaRt::useMart("ensembl", dataset = "hsapiens_gene_ensembl")
# Load mouse ensembl attributes
mouse = biomaRt::useMart("ensembl", dataset = "mmusculus_gene_ensembl")

hsa2mmEnsembl <- function(x,mouse,human){
  # Link both datasets and retrieve mouse genes from the human genes
  genes.list = biomaRt::getLDS(attributes = c("hgnc_symbol"), filters = "hgnc_symbol", values = x , mart = human, attributesL = c("mgi_symbol","ensembl_gene_id"), martL = mouse, uniqueRows = T)
  # Get unique names of genes 
  mouse.gene.list <- unique(genes.list[, 3])
  return(mouse.gene.list)
}

# load data from Seurat package
data(cc.genes)
# convert to mouse ensembl IDs
cc.genes.mouse <- lapply(cc.genes, hsa2mmEnsembl, mouse,human)
save(cc.genes.mouse, file=savefile)
}

```

## A comparison between the gene lists from Cyclone and Seurat with the different SVs

Match the gene list used in mmu.pairs for Cyclone (G1=12052 paired genes, S=6459 paired genes, G2M=9981 paired genes) and Seurat (S=40 and G2M=52) with the long gene lists from different SVs. Erik unlisted the paired gene lists from Cyclone, and made them unique. This resulted in three gene lists of G1=622, S=631, and G2M=641. 

```{r}
#unlists the mmu.pairs and create unique lists of genes for G1, S and G2M
mmu_vec_an_G1 <- unique(unlist(mmu.pairs["G1"]))
mmu_vec_an_S <- unique(unlist(mmu.pairs["S"]))
mmu_vec_an_G2M <- unique(unlist(mmu.pairs["G2M"]))
#unlists the annotation used in Seurat
seurat_vec_an_S <- unique(unlist(cc.genes.mouse["s.genes"]))
seurat_vec_an_G2M <- unique(unlist(cc.genes.mouse["g2m.genes"]))

#Matching the gene lists using intersect()
#sv2: ks value 7.816618e-45
#cyclone
ccg1_match_sv2 <- sort(intersect(mmu_vec_an_G1, gene.list.long_buet_SV2))
ccs_match_sv2 <- sort(intersect(mmu_vec_an_S, gene.list.long_buet_SV2))
ccg2m_match_sv2 <- sort(intersect(mmu_vec_an_G2M, gene.list.long_buet_SV2))
#seurat
seus_match_sv2 <- sort(intersect(seurat_vec_an_S, gene.list.long_buet_SV2))
seug2m_match_sv2 <- sort(intersect(seurat_vec_an_G2M, gene.list.long_buet_SV2))

#sv3: ks value 3.484951e-12
#cyclone
ccg1_match_sv3 <- sort(intersect(mmu_vec_an_G1, gene.list.long_buet_SV3))
ccs_match_sv3 <- sort(intersect(mmu_vec_an_S, gene.list.long_buet_SV3))
ccg2m_match_sv3 <- sort(intersect(mmu_vec_an_G2M, gene.list.long_buet_SV3))
#seurat
seus_match_sv3 <- sort(intersect(seurat_vec_an_S, gene.list.long_buet_SV3))
seug2m_match_sv3 <- sort(intersect(seurat_vec_an_G2M, gene.list.long_buet_SV3))

#sv5 ks value 2.922879e-36
#cyclone
ccg1_match_sv5 <- sort(intersect(mmu_vec_an_G1, gene.list.long_buet_SV5))
ccs_match_sv5 <- sort(intersect(mmu_vec_an_S, gene.list.long_buet_SV5))
ccg2m_match_sv5 <- sort(intersect(mmu_vec_an_G2M, gene.list.long_buet_SV5))
#seurat
seus_match_sv5 <- sort(intersect(seurat_vec_an_S, gene.list.long_buet_SV5))
seug2m_match_sv5 <- sort(intersect(seurat_vec_an_G2M, gene.list.long_buet_SV5))

#tables the number of matched genes between the different gene lists and SVs
res_gene_match <- data.frame(g1_cyclone_SV2=length(ccg1_match_sv2), s_cyclone_SV2=length(ccs_match_sv2), g2m_cyclone_SV2=length(ccg2m_match_sv2), g1_cyclone_SV3=length(ccg1_match_sv3), s_cyclone_SV3=length(ccs_match_sv3), g2m_cyclone_SV3=length(ccg2m_match_sv3), g1_cyclone_SV5=length(ccg1_match_sv5), s_cyclone_SV5=length(ccs_match_sv5), g2m_cyclone_SV5=length(ccg2m_match_sv5), s_seurat_SV2=length(seus_match_sv2), g2m_seurat_SV2=length(seug2m_match_sv2), s_seurat_SV3=length(seus_match_sv3), g2m_seurat_SV3=length(seug2m_match_sv3), s_seurat_SV5=length(seus_match_sv5), g2m_seurat_SV5=length(seug2m_match_sv5)) 
table(res_gene_match)

#lists the matched G2M genes from cyclone and seurat for sv2, sv3, and sv5
list(ccg2m_match_sv2)
list(seug2m_match_sv2)
list(ccg2m_match_sv3)
list(seug2m_match_sv3)
list(ccg2m_match_sv5)
list(seug2m_match_sv5)

```
Here Erik presents the matched genes in the predicted G2M phase for three different SVs mapping against cyclone (631 genes in G2M) and Seurat (52 genes in G2M). 40 Cyclone hits and 20 Seurat hits for sv2, 7 Cyclone hits and 1 Seurat hit for sv3, 40 Cyclone hits and 10 Seurat hits for sv5. Interesting note: 17 G2M genes are shared between Cyclone and Seurat for sv2, 1 G2M gene is shared between Cyclone and Seurat for sv3, 8 G2M genes are shared between Cyclone and Seurat for sv5. Similar comparison can be done to reduce the gene lists used in Cyclone and Seurat. Erik has not been able to test if reduced gene lists result in better prediction.