---
title: "Detecting cell-cycle stage difference in glioblastoma cells"
author: "Donghyung Lee"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{"Detecting cell-cycle stage difference in glioblastoma cells"}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


Here, we illustrate how to use the iasva package to detect cell cycle stage
difference within single cell RNA sequencing data. We use single cell RNA
sequencing (scRNA-Seq) data obtained from human glioblastoma samples
([Petel et. al., 2014](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4123637/)).
This dataset is included in a R data package ("iasvaExamples") containing data
examples for IA-SVA (https://github.com/dleelab/iasvaExamples). 
To install the package, follow the instruction provided in the GitHub page.

## Install packages
```{r install_packages, echo=TRUE, eval=FALSE}
#devtools
library(devtools)
#iasva
devtools::install_github("UcarLab/iasva")
#iasvaExamples  
devtools::install_github("dleelab/iasvaExamples")
```

## Load packages
```{r load_packages, echo=TRUE, message=FALSE}
rm(list=ls())
library(irlba) # partial SVD, the augmented implicitly restarted Lanczos bidiagonalization algorithm
library(iasva)
library(iasvaExamples)
library(sva)
library(SCnorm)
library(scran)
library(scater)
library(Rtsne)
library(pheatmap)
library(corrplot)
library(DescTools) #pcc i.e., Pearson's contingency coefficient
library(RColorBrewer)
library(SummarizedExperiment)
library(vioplot)
color.vec <- brewer.pal(3, "Set1")

# Normalization.
normalize <- function(counts) 
{
    normfactor <- colSums(counts)
    return(t(t(counts)/normfactor)*median(normfactor))
}
```

## Load the Buettner single cell RNA-Seq data
```{r load_data, echo=TRUE}
buettner.data <- readRDS("/Users/erikh/Desktop/git/scrna-cell-cycle/embryonic_data/EMTAB2805.rds")

# get tpms at gene level
buet_exprs <- assays(experiments(buettner.data)[["gene"]])[["TPM"]]
dim(buet_exprs)
# remove the .1 etc from the ensembl gene name
gname <- rownames(buet_exprs)
ensname <- gsub("\\.\\d+","",gname)
rownames(buet_exprs) <- ensname

# remove spike-ins from expression table
ercc <- grep("ERCC",rownames(buet_exprs))
buet_exprs <- buet_exprs[-ercc,]

# metadata with cell cycle stage assignment
M <- as.data.frame(colData(buettner.data))
head(M)
dim(buet_exprs)
```

## Filter out clearly low quality cells from Buettner data

Clearly some outlier cells on negative PC1 with low number of genes with pca using full dataset.
Mainly S-phase cells, even if most S-phase cells have a high median expression.

```{r filter}
# filter out cells with low nDet
nDet <- colSums(buet_exprs>1)
hist(nDet,n=100)
# make a cutoff at 4000 genes, removes 16 cells
keep <- which(nDet>4000)

buet_exprs<-buet_exprs[,keep]
M <- M[keep,]
dim(buet_exprs)

```

## Normalization for Buettner data
```{r normalization, fig.width=8, fig.height=6}
## count-depth relationship for all genes
norm_buet <- SingleCellExperiment(assays=list(counts = buet_exprs, logcounts = log2(buet_exprs+1)))
dim(norm_buet)
```

## Calculate the number of detected genes 
It is well known that the number of detected genes in each cell explains
a very large portion of variability in scRNA-Seq data 
([Hicks et. al. 2015 BioRxiv](http://biorxiv.org/content/early/2015/08/25/025528),
[McDavid et. al. 2016 Nature Biotechnology](http://www.nature.com/nbt/journal/v34/n6/full/nbt.3498.html)).
Frequently, the first principal component of log-transformed scRNA-Seq read counts is highly correlated
with the number of detected genes (e.g., r > 0.9). Here, we calculate the number of detected genes 
for glioblastoma cells, which will be used as an known factor in the IA-SVA analyses. 
```{r num_detected_genes, echo=TRUE, fig.width=7, fig.height=4}
Num_Detected_Genes_buet <- colSums(buet_exprs>0)
Geo_Lib_buet <- colSums(log(buet_exprs+1))
summary(Geo_Lib_buet)
barplot(Geo_Lib_buet, xlab="Cell", las=2, ylab = "Geometric library size")
lcounts_buet <- log(buet_exprs + 1)
# PC1 and Geometric library size correlation
pc1_buet = irlba(lcounts_buet - rowMeans(lcounts_buet), 1)$v[,1] ## partial SVD
cor(Num_Detected_Genes_buet, pc1_buet)
cor(Geo_Lib_buet, pc1_buet)
```

## Run IA-SVA
Here, we run IA-SVA using Geo_Lib_Size as a known factor and identify five hidden factors. 
SVs are plotted in a pairwise fashion to uncover which SVs can seperate cells. 
```{r run_iasva, echo=TRUE, fig.width= 7, fig.height=6}
set.seed(3445)
mod_buet <- model.matrix(~Geo_Lib_buet)
summ_exp_buet <- SummarizedExperiment(assays = buet_exprs)
iasva.res_buet <- iasva(summ_exp_buet, as.matrix(mod_buet[,-1]),verbose=FALSE, permute=FALSE, num.sv=7)
iasva.sv_buet <- iasva.res_buet$sv

Cluster_buet_7 <- as.factor(iasva.sv_buet[,7] < 0.1) 
levels(Cluster_buet_7)=c("Cluster1","Cluster2")
table(Cluster_buet_7)

pairs(iasva.sv_buet[,1:7], main="IA-SVA_7", pch=21, col=color.vec[Cluster_buet_7],
      bg=color.vec[Cluster_buet_7], oma=c(4,4,6,14))
legend("right", levels(Cluster_buet_7), fill=color.vec, bty="n")

plot(iasva.sv_buet[,c(1,7)], main="IA-SVA", pch=21, xlab="SV1", ylab="SV7",
     col=color.vec[Cluster_buet_7], bg=color.vec[Cluster_buet_7])

Cluster_buet_6 <- as.factor(iasva.sv_buet[,6] < 0.1) 
levels(Cluster_buet_6)=c("Cluster1","Cluster2")
table(Cluster_buet_6)

pairs(iasva.sv_buet[,1:7], main="IA-SVA_6", pch=21, col=color.vec[Cluster_buet_6],
      bg=color.vec[Cluster_buet_6], oma=c(4,4,6,14))
legend("right", levels(Cluster_buet_6), fill=color.vec, bty="n")

plot(iasva.sv_buet[,c(1,6)], main="IA-SVA", pch=21, xlab="SV1", ylab="SV6",
     col=color.vec[Cluster_buet_6], bg=color.vec[Cluster_buet_6])

Cluster_buet_5 <- as.factor(iasva.sv_buet[,5] < 0.1) 
levels(Cluster_buet_5)=c("Cluster1","Cluster2")
table(Cluster_buet_5)

pairs(iasva.sv_buet[,1:7], main="IA-SVA_5", pch=21, col=color.vec[Cluster_buet_5],
      bg=color.vec[Cluster_buet_5], oma=c(4,4,6,14))
legend("right", levels(Cluster_buet_5), fill=color.vec, bty="n")

plot(iasva.sv_buet[,c(1,5)], main="IA-SVA", pch=21, xlab="SV1", ylab="SV5",
     col=color.vec[Cluster_buet_5], bg=color.vec[Cluster_buet_5])

Cluster_buet_4 <- as.factor(iasva.sv_buet[,4] < 0.1) 
levels(Cluster_buet_4)=c("Cluster1","Cluster2")
table(Cluster_buet_4)

pairs(iasva.sv_buet[,1:7], main="IA-SVA_4", pch=21, col=color.vec[Cluster_buet_4],
      bg=color.vec[Cluster_buet_4], oma=c(4,4,6,14))
legend("right", levels(Cluster_buet_4), fill=color.vec, bty="n")

plot(iasva.sv_buet[,c(1,4)], main="IA-SVA", pch=21, xlab="SV1", ylab="SV4",
     col=color.vec[Cluster_buet_4], bg=color.vec[Cluster_buet_4])

Cluster_buet_3 <- as.factor(iasva.sv_buet[,3] < 0.1) 
levels(Cluster_buet_3)=c("Cluster1","Cluster2")
table(Cluster_buet_3)

pairs(iasva.sv_buet[,1:7], main="IA-SVA_3", pch=21, col=color.vec[Cluster_buet_3],
      bg=color.vec[Cluster_buet_3], oma=c(4,4,6,14))
legend("right", levels(Cluster_buet_3), fill=color.vec, bty="n")

plot(iasva.sv_buet[,c(1,3)], main="IA-SVA", pch=21, xlab="SV1", ylab="SV3",
     col=color.vec[Cluster_buet_3], bg=color.vec[Cluster_buet_3])

Cluster_buet_2 <- as.factor(iasva.sv_buet[,2] < 0.1) 
levels(Cluster_buet_2)=c("Cluster1","Cluster2")
table(Cluster_buet_2)

pairs(iasva.sv_buet[,1:7], main="IA-SVA_2", pch=21, col=color.vec[Cluster_buet_2],
      bg=color.vec[Cluster_buet_2], oma=c(4,4,6,14))
legend("right", levels(Cluster_buet_2), fill=color.vec, bty="n")

plot(iasva.sv_buet[,c(1,2)], main="IA-SVA", pch=21, xlab="SV1", ylab="SV2",
     col=color.vec[Cluster_buet_2], bg=color.vec[Cluster_buet_2])

Cluster_buet_1 <- as.factor(iasva.sv_buet[,1] < 0.1) 
levels(Cluster_buet_1)=c("Cluster1","Cluster2")
table(Cluster_buet_1)

pairs(iasva.sv_buet[,1:7], main="IA-SVA_1", pch=21, col=color.vec[Cluster_buet_1],
      bg=color.vec[Cluster_buet_1], oma=c(4,4,6,14))
legend("right", levels(Cluster_buet_1), fill=color.vec, bty="n")

plot(iasva.sv_buet[,c(1,2)], main="IA-SVA", pch=21, xlab="SV1", ylab="SV2",
     col=color.vec[Cluster_buet_1], bg=color.vec[Cluster_buet_1])

cor(Num_Detected_Genes_buet, iasva.sv_buet[,1:7])
cor(Geo_Lib_buet, iasva.sv_buet[,1:7])
corrplot(cor(iasva.sv_buet))
```
## Cell type assignment using scran R package
```{r run_scran, echo=TRUE, fig.width=6, fig.height=4}
mmu.pairs <- readRDS(system.file("exdata", 
                                "mouse_cycle_markers.rds", package="scran"))
assigned_buet <- cyclone(norm_buet, pairs=mmu.pairs)
head(assigned_buet$scores)
table(assigned_buet$phases)
phase_buet <- rep("S", ncol(norm_buet))
phase_buet[assigned_buet$scores$G1 > 0.5 & assigned_buet$scores$G2M < 0.5] <- "G1"
phase_buet[assigned_buet$scores$G1 < 0.5 & assigned_buet$scores$G2M > 0.5] <- "G2M"
phase_buet[assigned_buet$scores$G1 < 0.5 & assigned_buet$scores$G2M < 0.5] <- "S"
phase_buet[assigned_buet$scores$G1 > 0.5 & assigned_buet$scores$G2M > 0.5] <- "Unknown"
table(phase_buet)

G1_buet_SV <- iasva.sv_buet[phase_buet=="G1",]
S_buet_SV <- iasva.sv_buet[phase_buet=="S",]
G2M_buet_SV <- iasva.sv_buet[phase_buet=="G2M",]
Unknown_buet_SV <- iasva.sv_buet[phase_buet=="Unknown",]

corrplot(cor(G1_buet_SV))
corrplot(cor(S_buet_SV))
corrplot(cor(G2M_buet_SV))

vioplot(G1_buet_SV[,c(1)], S_buet_SV[,c(1)], G2M_buet_SV[,c(1)], names=c("G1", "S", "G2M"), col="gold")
title(xlab="Cell-cycle stage predictions", ylab="IA-SVA factor (SV1)")
vioplot(G1_buet_SV[,c(2)], S_buet_SV[,c(2)], G2M_buet_SV[,c(2)], names=c("G1", "S", "G2M"), 
   col="gold")
title(xlab="Cell-cycle stage predictions", ylab="IA-SVA factor (SV2)")
vioplot(G1_buet_SV[,c(3)], S_buet_SV[,c(3)], G2M_buet_SV[,c(3)], names=c("G1", "S", "G2M"), 
   col="gold")
title(xlab="Cell-cycle stage predictions", ylab="IA-SVA factor (SV3)")
vioplot(G1_buet_SV[,c(4)], S_buet_SV[,c(4)], G2M_buet_SV[,c(4)], names=c("G1", "S", "G2M"), 
   col="gold")
title(xlab="Cell-cycle stage predictions", ylab="IA-SVA factor (SV4)")
vioplot(G1_buet_SV[,c(5)], S_buet_SV[,c(5)], G2M_buet_SV[,c(5)], names=c("G1", "S", "G2M"), 
   col="gold")
title(xlab="Cell-cycle stage predictions", ylab="IA-SVA factor (SV5)")
vioplot(G1_buet_SV[,c(6)], S_buet_SV[,c(6)], G2M_buet_SV[,c(6)], names=c("G1", "S", "G2M"), 
   col="gold")
title(xlab="Cell-cycle stage predictions", ylab="IA-SVA factor (SV6)")
vioplot(G1_buet_SV[,c(7)], S_buet_SV[,c(7)], G2M_buet_SV[,c(7)], names=c("G1", "S", "G2M"), 
   col="gold")
title(xlab="Cell-cycle stage predictions", ylab="IA-SVA factor (SV7)")
```

## KS-test on the three cell cycle phases and the different SVs
The SV with the lowest KS value will be used in next section. The other SVs are ignored. The SV with lowest value should be the one best separating the cell cycle phases using the vioplot in previous section.
```{r }
#KS-test between G1 and G2M
ks_G1_G2M_1 <- ks.test(G1_buet_SV[,1],G2M_buet_SV[,1])
ks_G1_G2M_2 <- ks.test(G1_buet_SV[,2],G2M_buet_SV[,2])
ks_G1_G2M_3 <- ks.test(G1_buet_SV[,3],G2M_buet_SV[,3])
ks_G1_G2M_4 <- ks.test(G1_buet_SV[,4],G2M_buet_SV[,4])
ks_G1_G2M_5 <- ks.test(G1_buet_SV[,5],G2M_buet_SV[,5])
ks_G1_G2M_6 <- ks.test(G1_buet_SV[,6],G2M_buet_SV[,6])
ks_G1_G2M_7 <- ks.test(G1_buet_SV[,7],G2M_buet_SV[,7])
#Put the p-value in a variable
a1 <- as.numeric(unlist(ks_G1_G2M_1[2]))
a2 <- as.numeric(unlist(ks_G1_G2M_2[2]))
a3 <- as.numeric(unlist(ks_G1_G2M_3[2]))
a4 <- as.numeric(unlist(ks_G1_G2M_4[2]))
a5 <- as.numeric(unlist(ks_G1_G2M_5[2]))
a6 <- as.numeric(unlist(ks_G1_G2M_6[2]))
a7 <- as.numeric(unlist(ks_G1_G2M_7[2]))

#KS-test between G1 and S
ks_G1_S_1 <- ks.test(G1_buet_SV[,1],S_buet_SV[,1])
ks_G1_S_2 <- ks.test(G1_buet_SV[,2],S_buet_SV[,2])
ks_G1_S_3 <- ks.test(G1_buet_SV[,3],S_buet_SV[,3])
ks_G1_S_4 <- ks.test(G1_buet_SV[,4],S_buet_SV[,4])
ks_G1_S_5 <- ks.test(G1_buet_SV[,5],S_buet_SV[,5])
ks_G1_S_6 <- ks.test(G1_buet_SV[,6],S_buet_SV[,6])
ks_G1_S_7 <- ks.test(G1_buet_SV[,7],S_buet_SV[,7])
#Put the p-value in a variable
b1 <- as.numeric(unlist(ks_G1_S_1[2]))
b2 <- as.numeric(unlist(ks_G1_S_2[2]))
b3 <- as.numeric(unlist(ks_G1_S_3[2]))
b4 <- as.numeric(unlist(ks_G1_S_4[2]))
b5 <- as.numeric(unlist(ks_G1_S_5[2]))
b6 <- as.numeric(unlist(ks_G1_S_6[2]))
b7 <- as.numeric(unlist(ks_G1_S_7[2]))

#KS-test between S and G2M
ks_G2M_S_1 <- ks.test(G2M_buet_SV[,1],S_buet_SV[,1])
ks_G2M_S_2 <- ks.test(G2M_buet_SV[,2],S_buet_SV[,2])
ks_G2M_S_3 <- ks.test(G2M_buet_SV[,3],S_buet_SV[,3])
ks_G2M_S_4 <- ks.test(G2M_buet_SV[,4],S_buet_SV[,4])
ks_G2M_S_5 <- ks.test(G2M_buet_SV[,5],S_buet_SV[,5])
ks_G2M_S_6 <- ks.test(G2M_buet_SV[,6],S_buet_SV[,6])
ks_G2M_S_7 <- ks.test(G2M_buet_SV[,7],S_buet_SV[,7])
#Put the p-value in a variable
c1 <- as.numeric(unlist(ks_G2M_S_1[2]))
c2 <- as.numeric(unlist(ks_G2M_S_2[2]))
c3 <- as.numeric(unlist(ks_G2M_S_3[2]))
c4 <- as.numeric(unlist(ks_G2M_S_4[2]))
c5 <- as.numeric(unlist(ks_G2M_S_5[2]))
c6 <- as.numeric(unlist(ks_G2M_S_6[2]))
c7 <- as.numeric(unlist(ks_G2M_S_7[2]))

#Multiply the different KS-tests for every SV to get a feeling of which SVs to exclude.
ks_SV1 <- a1*b1*c1
ks_SV2 <- a2*b2*c2
ks_SV3 <- a3*b3*c3
ks_SV4 <- a4*b4*c4
ks_SV5 <- a5*b5*c5
ks_SV6 <- a6*b6*c6
ks_SV7 <- a7*b7*c7

```

## Find marker genes for the detected heterogeneity (SV2).
Here, using the find_markers() function we find marker genes that are significantly
associated with SV2 (multiple testing adjusted p-value < 0.05, 
default significance cutoff, and R-squared value > 0.4).
```{r find_markers_SV1, echo=TRUE, fig.width=6, fig.height=16}
# try different R2 thresholds
pdf(paste0("/Users/erikh/Desktop/git/scrna-cell-cycle/r_scripts/erik/ia_sva/sc_output/Clustering_analyses_figure3_sv2_buet.pdf"))
r2.results_buet <- study_R2(summ_exp_buet, iasva.sv_buet, selected.svs=2, no.clusters=2)
dev.off()

marker.counts.SV2_buet <- find_markers(summ_exp_buet, 
                            as.matrix(iasva.sv_buet[,c(2)]), rsq.cutoff = 0.45)
marker.counts.SV2.long_buet <- find_markers(summ_exp_buet, 
                              as.matrix(iasva.sv_buet[,c(2)]), rsq.cutoff = 0.4)
nrow(marker.counts.SV2_buet) 
nrow(marker.counts.SV2.long_buet)

anno.col2_buet <- data.frame(Cluster=Cluster_buet_2, SV1=iasva.sv_buet[,2])
rownames(anno.col2_buet) <- colnames(marker.counts.SV2_buet)
head(anno.col2_buet)

cluster.col_buet <- color.vec[1:2]
names(cluster.col_buet) <- as.vector(levels(Cluster_buet_2))
anno.colors_buet <- list(Cluster=cluster.col_buet)
anno.colors_buet

pheatmap(log(marker.counts.SV2_buet+1), show_colnames =FALSE, 
         clustering_method = "ward.D2", cutree_cols = 2, annotation_col = anno.col2_buet,
         annotation_colors = anno.colors_buet)

pheatmap(log(marker.counts.SV2.long_buet+1), show_colnames =FALSE, 
         clustering_method = "ward.D2", cutree_cols = 2, annotation_col = anno.col2_buet,
         annotation_colors = anno.colors_buet)

gene.list_buet_SV2 <- rownames(marker.counts.SV2_buet)
write.table(gene.list_buet_SV2, file = paste0("/Users/erikh/Desktop/git/scrna-cell-cycle/r_scripts/erik/ia_sva/sc_output/CC_genes.short_buet_SV2.txt"),
            col.names =F, row.names = F, quote = F)

gene.list.long_buet_SV2 <- rownames(marker.counts.SV2.long_buet)
write.table(gene.list.long_buet_SV2, file = paste0("/Users/erikh/Desktop/git/scrna-cell-cycle/r_scripts/erik/ia_sva/sc_output/CC_genes.long_buet_SV2.txt"),
            col.names =F, row.names = F, quote = F)

```

## Correlation between all SVs and the geometric library size
```{r SV1_geometric_lib_size, echo=TRUE}
cor(Num_Detected_Genes_buet, iasva.sv_buet[,c(1:7)])
cor(Geo_Lib_buet, iasva.sv_buet[,c(1:7)])
```